scvi/module/_multivae.py

```diff
@@ -0,0 +1,529 @@
+            x_expression, px_rate, px_r, px_dropout
+        )
+
+        # mix losses to get the correct loss for each cell
+        recon_loss = self._mix_modalities(
+            rl_accessibility + rl_expression,  # paired
+            rl_expression,  # expression
+            rl_accessibility,  # accessibility
+            mask_expr,
+            mask_acc,
+        )
+
+        # Compute KLD between Z and N(0,I)
+        qz_m = inference_outputs["qz_m"]
+        qz_v = inference_outputs["qz_v"]
+        kl_div_z = kld(
+            Normal(qz_m, torch.sqrt(qz_v)),
+            Normal(0, 1),
+        ).sum(dim=1)
+
+        # Compute KLD between distributions for paired data
+        qzm_expr = inference_outputs["qzm_expr"]
+        qzv_expr = inference_outputs["qzv_expr"]
+        qzm_acc = inference_outputs["qzm_acc"]
+        qzv_acc = inference_outputs["qzv_acc"]
+        kld_paired = kld(
+            Normal(qzm_expr, torch.sqrt(qzv_expr)), Normal(qzm_acc, torch.sqrt(qzv_acc))
+        ) + kld(
+            Normal(qzm_acc, torch.sqrt(qzv_acc)), Normal(qzm_expr, torch.sqrt(qzv_expr))
+        )
+        kld_paired = torch.where(
+            torch.logical_and(mask_acc, mask_expr),
+            kld_paired.T,
+            torch.zeros_like(kld_paired).T,
+        ).sum(dim=0)
+
+        # KL WARMUP
+        kl_local_for_warmup = kl_div_z
+        weighted_kl_local = kl_weight * kl_local_for_warmup
+
+        # PENALTY
+        # distance_penalty = kl_weight * torch.pow(z_acc - z_expr, 2).sum(dim=1)
+
+        # TOTAL LOSS
+        loss = torch.mean(recon_loss + weighted_kl_local + kld_paired)
+
+        kl_local = dict(kl_divergence_z=kl_div_z)
+        kl_global = torch.tensor(0.0)
+        return LossRecorder(loss, recon_loss, kl_local, kl_global)
+
+    def get_reconstruction_loss_expression(self, x, px_rate, px_r, px_dropout):
+        rl = 0.0
+        if self.gene_likelihood == "zinb":
+            rl = (
+                -ZeroInflatedNegativeBinomial(
+                    mu=px_rate, theta=px_r, zi_logits=px_dropout
+                )
+                .log_prob(x)
+                .sum(dim=-1)
+            )
+        elif self.gene_likelihood == "nb":
+            rl = -NegativeBinomial(mu=px_rate, theta=px_r).log_prob(x).sum(dim=-1)
+        elif self.gene_likelihood == "poisson":
+            rl = -Poisson(px_rate).log_prob(x).sum(dim=-1)
+        return rl
+
+    def get_reconstruction_loss_accessibility(self, x, p, d):
+        f = torch.sigmoid(self.region_factors) if self.region_factors is not None else 1
+        return torch.nn.BCELoss(reduction="none")(p * d * f, (x > 0).float()).sum(
+            dim=-1
+        )
+
+    @staticmethod
+    def _mix_modalities(x_paired, x_expr, x_acc, mask_expr, mask_acc):
+        x = torch.where(mask_expr.T, x_expr.T, x_acc.T).T
+        x = torch.where(torch.logical_and(mask_acc, mask_expr), x_paired.T, x.T).T
+        return x
```

Comment:
this is pretty close overall, @talashuach can you just add docstring to this function describing what's happening? I don't easily understand.